### PR TITLE
crimson/os/alienstore: improve alienstore's write parallelism

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
@liu-chunmei @tchaikov @athanatos 

I made a mistake yesterday, exclusive pipeline and crimson::os::AlienStore::tp_mutex do guarantee the order in which client requests are submitted to bluestore. The only problem is that the current tp_mutex is a AlienStore wide lock, while we only need a collection-scale lock. This pr is supposed to address that issue, please take a look, thanks:-)

File: src/crimson/os/alienstore/alien_collection.h [hunk 1]
```diff
@@ -19,8 +19,17 @@ class AlienCollection final : public FuturizedCollection {
 
   ~AlienCollection() {}
 
+  seastar::future<> lock_for_write() {
+    return trans_mutex.lock();
+  }
+
+  void unlock_for_write() {
+    trans_mutex.unlock();
+  }
```
Comment (tchaikov) [inline src/crimson/os/alienstore/alien_collection.h:28]:
```suggestion
  template <typename Func>
  seastar::future<> with_lock(Func&& func) {
    return seastar::with_lock(trans_mutex, std::move(func));
  }
```

File: src/crimson/os/alienstore/alien_collection.h [hunk 2]
```diff
@@ -19,8 +19,17 @@ class AlienCollection final : public FuturizedCollection {
...
+
 private:
   ObjectStore::CollectionHandle collection;
+  seastar::shared_mutex trans_mutex;
```
Comment (tchaikov) [inline src/crimson/os/alienstore/alien_collection.h:32]:
```suggestion
  seastar::shared_mutex mutex;
```

**Review by tchaikov — COMMENTED** [review]
> I made a mistake yesterday, exclusive pipeline and crimson::os::AlienStore::tp_mutex do guarantee the order in which client requests are submitted to bluestore. The only problem is that the current tp_mutex is a AlienStore wide lock, while we only need a collection-scale lock.

could you add more details on why this improves alienstore's parallelism when handling write ops? something like

> trade the grand per store tp_mutex for a finer grained per-collection lock for better concurrency

?

**Comment by rzarzynski** [comment]
@xxhdx1985126: nice! Have you taken a look on the performance impact in FIO / rados bench maybe?

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126: nice! Have you taken a look on the performance impact in FIO / rados bench maybe?

Will do so:-)

**Comment by markhpc** [comment]
Nice, if I have a chance I will take a look too.

**Comment by liu-chunmei** [comment]
@xxhdx1985126 Greate, we needn't guarantee the order of whole store, we only need guarantee the order in same pg. please do some test to see the performance.

**Comment by liu-chunmei** [comment]
@xxhdx1985126 maybe not much affect performance, since this mutex unlock quickly once  tp_submit return.  BTW, I am confused why you meet order issue, since the original tp_mutex has already guarantee the  strict sequence for same pg or not same pg.

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126 maybe not much affect performance, since this mutex unlock quickly once tp_submit return. 

Actually, I don't expect much performance improvement in cases where the underlying block device can serve little ops at one time, since, even if we can serve multiple client io ops simultaneously in AlienStore, the performance is still limited by the actualy disk. However, if the underlying device can serve multiple reqs concurrently, I think we may observe more evidences of performance improvement.

> BTW, I am confused why you meet order issue, since the original tp_mutex has already guarantee the strict sequence for same pg or not same pg.

I didn't actually met any order issues, I just wanted to make sure that the order is guaranteed. I think I can't rely on testing for make sure that as the issue can be hard to met in tests even if it does exists, so I chose to read the source code.

**Comment by xxhdx1985126** [comment]
@tchaikov Done:-)

File: src/crimson/os/alienstore/alien_store.cc [hunk 3]
```diff
@@ -363,17 +363,17 @@ seastar::future<> AlienStore::do_transaction(CollectionRef ch,
...
+	return alien_coll->with_lock([this, ch, id, &txn, &done] {
 	  Context *crimson_wrapper =
 	    ceph::os::Transaction::collect_all_contexts(txn);
 	  return tp->submit([this, ch, id, crimson_wrapper, &txn, &done] {
 	    txn.register_on_commit(new OnCommit(id, done, crimson_wrapper, txn));
 	    auto c = static_cast<AlienCollection*>(ch.get());
 	    return store->queue_transaction(c->collection, std::move(txn));
 	  });
-	}).then([this, &done] (int r) {
+	}).then([this, alien_coll, &done] (int r) {
```
Comment (tchaikov) [inline src/crimson/os/alienstore/alien_store.cc:375]: could drop `alien_coll`, which is not used anymore in the latest revision.

Comment (xxhdx1985126) [reply src/crimson/os/alienstore/alien_store.cc]: @tchaikov Sorry for the inconvenience, just corrected it:-)

Comment (tchaikov) [reply src/crimson/os/alienstore/alien_store.cc]: no worries. thank you!

**Review by tchaikov — APPROVED** [review]

**Pull request merged**